concordium-consensus/src/Concordium/GlobalState/Persistent/Account/StructureV1.hs

```diff
@@ -0,0 +1,1314 @@
+    load = do
+        accountNonce <- get
+        accountAmount <- get
+        accountStakedAmount <- get
+        mEnduringData <- load
+        return $ do
+            accountEnduringData <- mEnduringData
+            return $! PersistentAccount{..}
+
+-- |Get the enduring data for an account.
+enduringData :: PersistentAccount av -> PersistentAccountEnduringData av
+enduringData = eagerBufferedDeref . accountEnduringData
+
+-- |Get the persisting data for an account.
+persistingData :: PersistentAccount av -> PersistingAccountData
+persistingData = eagerBufferedDeref . paedPersistingData . enduringData
+
+-- ** Queries
+
+-- |Get the canonical address of the account.
+getCanonicalAddress :: (Monad m) => PersistentAccount av -> m AccountAddress
+getCanonicalAddress acc = do
+    let pd = persistingData acc
+    return $! pd ^. accountAddress
+
+-- |Get the current public account balance.
+getAmount :: (Monad m) => PersistentAccount av -> m Amount
+getAmount = pure . accountAmount
+
+-- |Gets the amount of a baker's stake, or 'Nothing' if the account is not a baker.
+getBakerStakeAmount :: (Monad m) => PersistentAccount av -> m (Maybe Amount)
+getBakerStakeAmount acc = do
+    let ed = enduringData acc
+    return $! case paedStake ed of
+      PersistentAccountStakeEnduringBaker{} -> Just $! accountStakedAmount acc
+      _ -> Nothing
+
+-- |Get the amount that is staked on the account.
+getStakedAmount :: (Monad m) => PersistentAccount av -> m Amount
+getStakedAmount acc = return $! accountStakedAmount acc
+
+-- |Get the amount that is locked in scheduled releases on the account.
+getLockedAmount :: (Monad m) => PersistentAccount av -> m Amount
+getLockedAmount acc = do
+    let ed = enduringData acc
+    return $! paedLockedAmount ed
+
+-- | Get the current public account available balance.
+-- This accounts for lock-up and staked amounts.
+-- @available = total - max locked staked@
+getAvailableAmount :: (Monad m) => PersistentAccount av -> m Amount
+getAvailableAmount acc = do
+    let ed = enduringData acc
+    return $! accountAmount acc - max (accountStakedAmount acc) (paedLockedAmount ed)
+
+-- |Get the next account nonce for transactions from this account.
+getNonce :: (Monad m) => PersistentAccount av -> m Nonce
+getNonce = pure . accountNonce
+
+-- |Determine if a given operation is permitted for the account.
+--
+-- * For 'AllowedEncryptedTransfers' the account may only have 1 credential.
+--
+-- * For 'AllowedMultipleCredentials' the account must have the empty encrypted balance.
+isAllowed :: Monad m => PersistentAccount av -> AccountAllowance -> m Bool
+isAllowed acc AllowedEncryptedTransfers = do
+    creds <- getCredentials acc
+    return $! Map.size creds == 1
+isAllowed acc AllowedMultipleCredentials = do
+    let ed = enduringData acc
+    -- We use the invariant that if the encrypted amount is present then it will be non-empty.
+    return $! case paedEncryptedAmount ed of
+      Null -> True
+      Some _ -> False
```

Comment:
Since we are changing the protocol then we don't have to keep the semantics exactly, but it seems to me that the semantics here has changed a bit. Previously if you ended up with 0 (encrypted in a specific way) you could then add multiple credentials on the account.

In this version it seems we never go back to `Null` once we do anything with encrypted transfers. This is probably OK, but it is a change in behaviour that was not planned.

Reply:
I have changed this back so it will do the `isZero` check now.